client.go

```diff
@@ -0,0 +1,77 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package cluster
+
+// KVStore provides access to the configuration store
+type KVStore interface {
+	// Get retrieves the value for the given key, marshalling it into the provided value
+	Get(key string, v interface{}) error
+
+	// Put stores the value for the given key
+	Put(key string, v interface{}) error
+}
+
+// A ServiceInstance is a single instance of a service
+type ServiceInstance struct {
```

Comment:
Can we perhaps make all the structs simply `interface`s with accessors e.g. "Service() string" and where mutable with setters too "SetHealth(fn func() error)"?

Have found that when you depend on a third party vendor "A" and they include struct "X" while another third party "B" itself depends on "A" which includes struct "X" it tends to cause a lot of issues.  This is why we couldn't use Bark in M3DB because one of our other dependencies had a dependency on Bark too and said that the structs in Bark were being declared multiple times.

Also everything being interfaces does make test relatively less painful as it's very easy to generate a mock.